ratis-common/src/main/java/org/apache/ratis/util/BatchLogger.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ratis.util;
+
+import org.apache.ratis.thirdparty.com.google.common.cache.Cache;
+import org.apache.ratis.thirdparty.com.google.common.cache.CacheBuilder;
+import org.apache.ratis.thirdparty.com.google.common.collect.Lists;
+import org.slf4j.Logger;
+
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+public final class BatchLogger {
+
+  private BatchLogger() {
+  }
+
+  public final static class UniqueId {
+    private final List<Object> components;
```

Comment:
How about using a `String` and a `StackTraceElement` (which has filename and line number)?
```java
  public final static class UniqueId {
    private final StackTraceElement traceElement;
    private final String name;

    private UniqueId(StackTraceElement traceElement, String name) {
      this.traceElement = traceElement;
      this.name = name;
    }

    @Override
    public boolean equals(Object obj) {
      if (obj == this) {
        return true;
      } else if (!(obj instanceof UniqueId)) {
        return false;
      }

      final UniqueId that = (UniqueId) obj;
      return Objects.equals(this.traceElement, that.traceElement)
          && Objects.equals(this.name, that.name);
    }

    @Override
    public int hashCode() {
      return traceElement.hashCode() ^ name.hashCode();
    }
  }
```

Reply:
That is more semantic friendly! Let me try if it works all the same.